File: py_codes/oo_design/monster_game/monster.py
class Monster:
	'''
	I defined map here as static variable,
	Because the monsters are all created on a single map
	'''
	_map = None
	@staticmethod 
	def setGameMap(game_map):
		Monster._map = game_map

	def __init__(self, monster_id):
		self._id = monster_id
		self._alive = True
		self._city_name = None
		if not Monster._map:
			print "Static map object of monster must be set before initialization!"
			return
		Monster._map.placeMonsterRandomly(self)

	def getId(self):
		return self._id

	def getCityName(self):
		return self._city_name

	def setCityName(self, city_name):
		self._city_name = city_name

	def kill(self):
		self._alive = False

	def isAlive(self):
		return self._alive

	def move(self):
		Monster._map.moveMonster(self)

File: c_codes/oo_design/monster_game/monster.cpp
#include "monster.h"


Monster::Monster()
{
	_city = NULL;
	_game_map = NULL;
	_id = -1;
	_alive = true;
	_trapped = false;
}


void Monster::setId(int monster_id)
{
	_id = monster_id;
}


void Monster::setGameMap(GameMap * map)
{	
	_game_map = map;
}


int Monster::getId()
{
	return _id;
}


City* Monster::getCity()
{
	return _city;
}


bool Monster::isAlive()
{
	return _alive;
}


void Monster::kill()
{
	_alive = false;
}


bool Monster::isTrapped()
{
	return _trapped;
}


void Monster::placeRandomly()
{
	// If the initial placement is not successful, set the state as "trapped"
	_city = _game_map->placeMonsterRandomly();
	if(! _city)
	{
		_trapped = true;
	}
}


void Monster::move()
{
	City *next_city;

	// If no next city is available, set monster as "trapped"
	next_city = _game_map->moveMonster(_city);
	if(next_city)
	{
		_city = next_city;
	}
	else
	{
		_trapped = true;
	}
}